packages/sanity/src/_internal/cli/commands/backup/downloadBackupCommand.ts

```diff
@@ -0,0 +1,288 @@
+    const docOutStream = createWriteStream(tmpOutDocumentsFile)
+    const docWriteMutex = new Mutex()
+
+    try {
+      const backupFileStream = new PaginatedGetBackupStream(
+        client,
+        opts.projectId,
+        opts.datasetName,
+        opts.backupId,
+        opts.token,
+      )
+
+      const files: File[] = []
+      let i = 0
+      for await (const file of backupFileStream) {
+        files.push(file)
+        i++
+        progressSpinner.set({
+          step: `Reading backup files...`,
+          update: true,
+          current: i,
+          total: backupFileStream.totalFiles,
+        })
+      }
+
+      let totalItemsDownloaded = 0
+      // This is dynamically imported because this module is ESM only and this file gets compiled to CJS at this time.
+      const {default: pMap} = await import('p-map')
+      await pMap(
+        files,
+        async (file: File) => {
+          if (file.type === 'file' || file.type === 'image') {
+            await downloadAsset(file.url, file.name, file.type, tmpOutDir)
+          } else {
+            const doc = await downloadDocument(file.url)
+            await docWriteMutex.runExclusive(() => {
+              docOutStream.write(`${doc}\n`)
+            })
+          }
+
+          totalItemsDownloaded += 1
+          progressSpinner.set({
+            step: `Downloading documents and assets...`,
+            update: true,
+            current: totalItemsDownloaded,
+            total: backupFileStream.totalFiles,
+          })
+        },
+        {concurrency: opts.concurrency},
+      )
+    } catch (error) {
+      progressSpinner.fail()
+      let msg = error.statusCode ? error.response.body.message : error.message
+      // If no message can be extracted, print the whole error.
+      if (msg === undefined) {
+        msg = String(error)
+      }
+      throw new Error(`Downloading dataset backup failed: ${msg}`)
+    }
+
+    progressSpinner.set({step: `Archiving files into a tarball...`, update: true})
+    try {
+      await archiveDir(tmpOutDir, outFilePath, (processedBytes: number) => {
+        progressSpinner.update({
+          step: `Archiving files into a tarball, ${humanFileSize(processedBytes)} bytes written...`,
+        })
+      })
+    } catch (err) {
+      progressSpinner.fail()
+      throw new Error(`Archiving backup failed: ${err.message}`)
+    }
+
+    progressSpinner.set({
+      step: `Cleaning up temporary files at ${chalk.cyan(`${tmpOutDir}`)}`,
+    })
+    cleanupTmpDir(tmpOutDir)
+
+    progressSpinner.set({
+      step: `Backup download complete [${prettyMs(Date.now() - start)}]`,
+    })
+    progressSpinner.succeed()
+  },
+}
+
+// prepareBackupOptions validates backup options from CLI and prepares Client and DownloadBackupOptions.
+async function prepareBackupOptions(
+  context: CliCommandContext,
+  args: CliCommandArguments,
+): Promise<[SanityClient, DownloadBackupOptions]> {
+  const flags = args.extOptions
```

Comment:
Take a look at this [PR](https://github.com/sanity-io/sanity/pull/5616/files#diff-5dd59c387c299ba5b1746c4c0a69aed0ec8a759c264d799e8462f6a03f7e4a91R68) I believe we want to use `yargs` to parse the flags, it enforces types and in future we want to move to making sure all flags are defined properly. It will also help remove some of the type checking code below

Reply:
Good point. Updated code to follow this pattern.